frontend/controller/service-worker.js

```diff
@@ -246,20 +171,86 @@ sbp('sbp/selectors/register', {
   }
 })
 
-// helper method
+// Events that need to be relayed to the SW
+;[LOGIN_COMPLETE, NEW_CHATROOM_UNREAD_POSITION, SET_APP_LOGS_FILTER].forEach((event) =>
+  sbp('okTurtles.events/on', event, (...data) => {
+    navigator.serviceWorker.controller?.postMessage({ type: 'event', subtype: event, data })
+  })
+)
 
-function urlBase64ToUint8Array (base64String) {
-  // reference: https://gist.github.com/Klerith/80abd742d726dd587f4bd5d6a0ab26b6
-  const padding = '='.repeat((4 - (base64String.length % 4)) % 4)
-  const base64 = (base64String + padding)
-    .replace(/-/g, '+')
-    .replace(/_/g, '/')
-
-  const rawData = window.atob(base64)
-  const outputArray = new Uint8Array(rawData.length)
+const swRpc = (() => {
+  if (!navigator.serviceWorker) {
+    throw new Error('Missing service worker object')
+  }
+  let controller: ?ServiceWorker = navigator.serviceWorker.controller
+  navigator.serviceWorker.addEventListener('controllerchange', (ev: Event) => {
+    controller = (navigator.serviceWorker: any).controller
+  }, false)
+
+  return (...args) => {
+    let cleanup
+    return new Promise((resolve, reject) => {
+      if (!controller) {
+        reject(new Error('Service worker not ready'))
+        return
+      }
+      const messageChannel = new MessageChannel()
+      const onmessage = (event: MessageEvent) => {
+        if (event.data && Array.isArray(event.data)) {
+          const r = deserializer(event.data[1])
+          // $FlowFixMe[incompatible-use]
+          if (event.data[0] === true) {
+            resolve(r)
+          } else {
+            reject(r)
+          }
+          cleanup()
+        }
+      }
+      const onmessageerror = (event: MessageEvent) => {
+        reject(event.data)
+        cleanup()
+      }
+      cleanup = () => {
+        // This can help prevent memory leaks if the GC doesn't clean up once
+        // the port goes out of scope
+        messageChannel.port1.removeEventListener('message', onmessage, false)
+        messageChannel.port1.removeEventListener('messageerror', onmessageerror, false)
+        messageChannel.port1.close()
+        // Remove potential references to the messageChannel
+        cleanup = undefined
+      }
+      messageChannel.port1.addEventListener('message', onmessage, false)
+      messageChannel.port1.addEventListener('messageerror', onmessageerror, false)
+      messageChannel.port1.start()
+      const { data, transferables } = serializer(args)
+      controller.postMessage({
+        type: 'sbp',
+        port: messageChannel.port2,
+        data
+      }, [messageChannel.port2, ...transferables])
+    }).finally(cleanup)
```

Comment:
Does it make sense to define this `cleanup` variable outside of the Promise and call it here in `.finally`?

And is it safe to call `.finally` on `undefined`? Seems safe...

```
> let foo
undefined
> new Promise( (resolve, reject) => { resolve()  }).finally(foo)
Promise {
  <pending>,
  [Symbol(async_id_symbol)]: 487,
  [Symbol(trigger_async_id_symbol)]: 486
}
> new Promise( (resolve, reject) => { resolve()  }).finally(foo)
Promise {
  <pending>,
  [Symbol(async_id_symbol)]: 497,
  [Symbol(trigger_async_id_symbol)]: 496
}
```

Reply:
It makes it DRYer by avoiding the need to call `cleanup` in two locations (onmessage and onmessagerror).

As for the safety, it depends on whether `undefined` is considered an object (can't tell from the spec). But from testing it seems safe.

I'll update this to call cleanup in those two locations instead.